client/quota.go

```diff
@@ -0,0 +1,85 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2021 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package client
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+
+	"golang.org/x/xerrors"
+)
+
+type postQuotaData struct {
+	Action    string   `json:"action"`
+	GroupName string   `json:"group-name"`
+	Parent    string   `json:"parent,omitempty"`
+	Snaps     []string `json:"snaps,omitempty"`
+	MaxMemory uint64   `json:"max-memory"`
+}
+
+type QuotaGroupResult struct {
+	GroupName string   `json:"group-name"`
+	Parent    string   `json:"parent,omitempty"`
+	Subgroups []string `json:"subgroups,omitempty"`
+	Snaps     []string `json:"snaps,omitempty"`
+	MaxMemory uint64   `json:"max-memory"`
+}
+
+// EnsureQuota creates a quota group or updates an existing group.
+// The list of snaps can be empty.
+func (client *Client) EnsureQuota(groupName string, parent string, snaps []string, maxMemory uint64) error {
+	if groupName == "" {
+		return xerrors.Errorf("cannot create or update quota group without a name")
+	}
+	// TODO: use naming.ValidateQuotaGroup()
```

Comment:
(nitpick^2) given that the server side will validate I'm not sure we really need to client side validate but I'm happy to keep this todo